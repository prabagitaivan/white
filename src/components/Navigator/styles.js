import { makeStyles } from '@material-ui/styles'

const layout = 'Navigator'

export default makeStyles(
  {
    appBar: {
      height: 50,
      justifyContent: 'center',
      top: ({ desktop }) => (desktop ? 0 : 'auto'),
      bottom: ({ desktop }) => (desktop ? 'auto' : 0)
    },
    grow: {
      flexGrow: 1
    },
    fab: {
      position: 'absolute',
      zIndex: 1,
      top: ({ desktop }) => (desktop ? 'initial' : -25),
      bottom: ({ desktop }) => (desktop ? -25 : 'initial'),
      left: 0,
      right: 0,
      margin: '0 auto'
    },
    emoji: {
      position: 'absolute',
      top: 6.5,
      bottom: 0,
      left: 0,
      right: 0,
      fontSize: 30
    },
    compressIcon: {
      padding: 2
    },
    themeContainer: {
      marginLeft: 10
    }
  },
  { name: layout }
)

export const menu = makeStyles(
  {
    button: {
      fontSize: ({ desktop }) => (desktop ? 12 : 10),
      textTransform: 'initial',

      '& .MuiButton-startIcon': {
        marginTop: -3,
        marginRight: 2
      },
      '& .MuiButton-startIcon > svg': {
        fontSize: 24
      }
    },
    menu: {
      position: 'fixed',
      top: ({ desktop }) => (desktop ? 55 : 'initial'),
      left: ({ desktop }) => (desktop ? 17 : 7),
      bottom: ({ desktop }) => (desktop ? 'initial' : 55)
    },
    option: {
      paddingLeft: 8,
      paddingRight: 8,

      '& .MuiButton-root': {
        fontSize: ({ desktop }) => (desktop ? 12 : 10),
        textTransform: 'initial',
        backgroundColor: 'transparent'
      },
      '& .MuiButton-startIcon': {
        marginRight: 2
      },
      '& .MuiButton-startIcon > svg': {
        fontSize: 24
      }
    }
  },
  { name: `${layout}Menu` }
)
