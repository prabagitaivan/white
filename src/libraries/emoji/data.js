const data = [
  { image: '😁', text: 'grinning face with smiling eyes' },
  { image: '😂', text: 'face with tears of joy' },
  { image: '🙂', text: 'slightly smiling face' },
  { image: '🙃', text: 'upside-down face' },
  { image: '🤔', text: 'thinking face' },
  { image: '🤨', text: 'face with raised eyebrow' },
  { image: '🤯', text: 'exploding head' },
  { image: '😮', text: 'face with open mouth' },
  { image: '👻', text: 'ghost' },
  { image: '👽', text: 'alien' },
  { image: '👾', text: 'alien monster' },
  { image: '🤖', text: 'robot' },
  { image: '🙈', text: 'see-no-evil monkey' },
  { image: '🙉', text: 'hear-no-evil monkey' },
  { image: '🙊', text: 'speak-no-evil monkey' },
  { image: '👋', text: 'waving hand' },
  { image: '🖖', text: 'vulcan salute' },
  { image: '✌️', text: 'victory hand' },
  { image: '🤟', text: 'love-you gesture' },
  { image: '🤘', text: 'sign of the horns' },
  { image: '👍', text: 'thumbs up' },
  { image: '👊', text: 'oncoming fist' },
  { image: '🚀', text: 'rocket' },
  { image: '🔥', text: 'fire' }
]

export default data
