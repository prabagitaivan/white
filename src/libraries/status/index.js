export const isLightMode = () => {
  const hour = new Date().getHours()

  if (hour >= 6 && hour <= 17) return true
  else return false
}

export const isMacPlatform = () => {
  // use deprecated version as the new one still have minim browser support
  // https://developer.mozilla.org/en-US/docs/Web/API/NavigatorUAData/platform
  const platform = window.navigator.platform || ''

  return platform.indexOf('Mac') === 0
}
