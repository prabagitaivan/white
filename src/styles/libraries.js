export const generateClassName = (rule, styleSheet) => {
  const prefix = styleSheet.options.classNamePrefix
  const name = `${prefix}-${rule.key}`

  if (prefix.match(/^Mui/)) return name
  else return `White${name}`
}

export const getLightStatus = () => {
  const hour = new Date().getHours()

  if (hour >= 6 && hour <= 17) return true
  else return false
}
