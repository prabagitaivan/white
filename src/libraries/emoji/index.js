import data from './data'

export const getRandomEmoji = () => {
  const limit = data.length - 1
  const index = Math.floor(Math.random() * limit)

  return data[index]
}
