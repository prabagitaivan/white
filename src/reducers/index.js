import { combineReducers } from 'redux'
import status from './status'
import randomNotes from './randomNotes'
import treeBookmarks from './treeBookmarks'

const reducers = combineReducers({
  status,
  randomNotes,
  treeBookmarks
})

export default reducers
