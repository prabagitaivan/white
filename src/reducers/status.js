import { createActions, handleActions } from 'redux-actions'
import { getLightStatus } from '../libraries/styles'

const defaultState = {
  online: true,
  desktop: true,
  light: getLightStatus(),
  page: 'RandomNotes'
}

export const {
  status: { setConnection, setScreen, setTheme, setPage }
} = createActions({
  STATUS: {
    SET_CONNECTION: action => action,
    SET_SCREEN: action => action,
    SET_THEME: action => action,
    SET_PAGE: action => action
  }
})

export default handleActions(
  {
    [setConnection]: (state, { payload: online }) => ({ ...state, online }),
    [setScreen]: (state, { payload: size }) => ({
      ...state,
      desktop: size > 600
    }),
    [setTheme]: (state, { payload: light }) => ({ ...state, light }),
    [setPage]: (state, { payload: page }) => ({ ...state, page })
  },
  defaultState
)
