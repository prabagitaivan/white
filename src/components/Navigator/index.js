import React, { memo, useState } from 'react'
import { useSelector, useDispatch } from 'react-redux'
import {
  AppBar,
  Fab,
  Avatar,
  Container,
  Toolbar,
  Tooltip,
  IconButton
} from '@material-ui/core'
import {
  PlayCircleFilled,
  GitHub,
  WbSunny,
  Brightness2
} from '@material-ui/icons'
import useStyles from './styles'
import { setTheme } from '../../reducers/status'
import { getRandomEmoji } from '../../libraries/emoji'

const initialEmoji = getRandomEmoji()

const openRepository = () =>
  window.open('https://github.com/prabagitaivan/white')

export default memo(() => {
  const [emoji, setEmoji] = useState(initialEmoji)
  const { desktop, light } = useSelector(state => state.status)
  const dispatch = useDispatch()
  const classes = useStyles({ desktop })

  const changeTheme = () => {
    dispatch(setTheme(!light))
  }
  const changeEmoji = () => {
    const emoji = getRandomEmoji()
    setEmoji(emoji)
  }

  return (
    <AppBar id='navigator' position='fixed' className={classes.appBar}>
      <Fab color='secondary' className={classes.fab} onClick={changeEmoji}>
        <Avatar>
          <span role='img' aria-label={emoji.text} className={classes.emoji}>
            {emoji.image}
          </span>
        </Avatar>
      </Fab>
      <Container>
        <Toolbar disableGutters>
          Random Notes
          <div className={classes.grow} />
          <Tooltip title='Random Notes'>
            <IconButton size='small'>
              <PlayCircleFilled />
            </IconButton>
          </Tooltip>
          <Tooltip title='Repository'>
            <IconButton size='small' onClick={openRepository}>
              <GitHub className={classes.compressIcon} />
            </IconButton>
          </Tooltip>
          <Tooltip
            title='Light / Dark Theme'
            edge='end'
            className={classes.themeContainer}
          >
            <IconButton size='small' onClick={changeTheme}>
              {light ? (
                <WbSunny className={classes.compressIcon} />
              ) : (
                <Brightness2 className={classes.compressIcon} />
              )}
            </IconButton>
          </Tooltip>
        </Toolbar>
      </Container>
    </AppBar>
  )
})
