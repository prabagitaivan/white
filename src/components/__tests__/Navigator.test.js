import { render } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import React from 'react'
import { Provider } from 'react-redux'
import createStore from '../../stores'
import Styles from '../../styles'
import { setTheme } from '../../reducers/status'
import * as emoji from '../../libraries/emoji'
import Navigator from '../Navigator'

let store

function renderNavigator ({ desktop = true, light = true }) {
  const preloadedState = { status: { desktop, light } }
  store = createStore(preloadedState)

  jest.spyOn(store, 'dispatch').mockReturnValue()

  return render(
    <Provider store={store}>
      <Styles>
        <Navigator />
      </Styles>
    </Provider>
  )
}

describe('components Navigator', () => {
  describe('snapshots', () => {
    it('contain navigations and its icon on light theme', () => {
      const { getByText, getByTitle } = renderNavigator({})
      expect(getByText('Random Notes')).toBeInTheDocument()
      expect(getByTitle('Random Notes')).toMatchSnapshot()
      expect(getByTitle('Repository')).toMatchSnapshot()
      expect(getByTitle('Light / Dark Theme')).toMatchSnapshot()
    })
    it('contain navigations and its icon on dark theme', () => {
      const { getByTitle } = renderNavigator({ light: false })
      expect(getByTitle('Light / Dark Theme')).toMatchSnapshot()
    })
    it('show in top for desktop', () => {
      const { container } = renderNavigator({})
      expect(container.firstChild).toHaveStyle('top: 0px')
    })
    it('show in bottom for mobile', () => {
      const { container } = renderNavigator({ desktop: false })
      expect(container.firstChild).toHaveStyle('bottom: 0px')
    })
  })
  describe('userEvent', () => {
    beforeEach(() => {
      jest.spyOn(window, 'open').mockReturnValue()
      jest.spyOn(emoji, 'getRandomEmoji')
    })
    afterEach(() => {
      window.open.mockRestore()
      emoji.getRandomEmoji.mockRestore()
      store.dispatch.mockRestore()
    })

    it('open this repository when click the Repository', () => {
      const { getByTitle } = renderNavigator({})
      const repo = getByTitle('Repository')

      userEvent.click(repo)
      expect(window.open).toHaveBeenCalledWith(
        'https://github.com/prabagitaivan/white'
      )
    })
    it('random emoji when click the Emoji avatar', () => {
      const { getByRole } = renderNavigator({})
      const avatar = getByRole('img')

      userEvent.click(avatar)
      expect(emoji.getRandomEmoji).toHaveBeenCalledTimes(1)

      userEvent.click(avatar)
      expect(emoji.getRandomEmoji).toHaveBeenCalledTimes(2)
    })
    it('switch to dark theme when click the Light / Dark Theme in light theme', () => {
      const { getByTitle } = renderNavigator({})
      const theme = getByTitle('Light / Dark Theme')

      userEvent.click(theme)
      expect(store.dispatch).toHaveBeenCalledTimes(1)
      expect(store.dispatch).toHaveBeenCalledWith(setTheme(false))
    })
    it('switch to light theme when click the Light / Dark Theme in dark theme', () => {
      const { getByTitle } = renderNavigator({ light: false })
      const theme = getByTitle('Light / Dark Theme')

      userEvent.click(theme)
      expect(store.dispatch).toHaveBeenCalledTimes(1)
      expect(store.dispatch).toHaveBeenCalledWith(setTheme(true))
    })
  })
})
