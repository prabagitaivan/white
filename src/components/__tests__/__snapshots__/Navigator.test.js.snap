// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components Navigator snapshots contain navigations and its icon on dark theme 1`] = `
<button
  class="MuiButtonBase-root MuiIconButton-root WhiteNavigator-themeContainer MuiIconButton-sizeSmall MuiIconButton-edgeEnd"
  tabindex="0"
  title="Light / Dark Theme"
  type="button"
>
  <span
    class="MuiIconButton-label"
  >
    <svg
      aria-hidden="true"
      class="MuiSvgIcon-root WhiteNavigator-compressIcon"
      focusable="false"
      role="presentation"
      viewBox="0 0 24 24"
    >
      <path
        d="M10 2c-1.82 0-3.53.5-5 1.35C7.99 5.08 10 8.3 10 12s-2.01 6.92-5 8.65C6.47 21.5 8.18 22 10 22c5.52 0 10-4.48 10-10S15.52 2 10 2z"
      />
    </svg>
  </span>
  <span
    class="MuiTouchRipple-root"
  />
</button>
`;

exports[`components Navigator snapshots contain navigations and its icon on light theme 1`] = `
<button
  class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeSmall"
  tabindex="0"
  title="Random Notes"
  type="button"
>
  <span
    class="MuiIconButton-label"
  >
    <svg
      aria-hidden="true"
      class="MuiSvgIcon-root"
      focusable="false"
      role="presentation"
      viewBox="0 0 24 24"
    >
      <path
        d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm-2 14.5v-9l6 4.5-6 4.5z"
      />
    </svg>
  </span>
  <span
    class="MuiTouchRipple-root"
  />
</button>
`;

exports[`components Navigator snapshots contain navigations and its icon on light theme 2`] = `
<button
  class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeSmall"
  tabindex="0"
  title="Repository"
  type="button"
>
  <span
    class="MuiIconButton-label"
  >
    <svg
      aria-hidden="true"
      class="MuiSvgIcon-root WhiteNavigator-compressIcon"
      focusable="false"
      role="presentation"
      viewBox="0 0 24 24"
    >
      <path
        d="M12 .3a12 12 0 0 0-3.8 23.4c.6.1.8-.3.8-.6v-2c-3.3.7-4-1.6-4-1.6-.6-1.4-1.4-1.8-1.4-1.8-1-.7.1-.7.1-.7 1.2 0 1.9 1.2 1.9 1.2 1 1.8 2.8 1.3 3.5 1 0-.8.4-1.3.7-1.6-2.7-.3-5.5-1.3-5.5-6 0-1.2.5-2.3 1.3-3.1-.2-.4-.6-1.6 0-3.2 0 0 1-.3 3.4 1.2a11.5 11.5 0 0 1 6 0c2.3-1.5 3.3-1.2 3.3-1.2.6 1.6.2 2.8 0 3.2.9.8 1.3 1.9 1.3 3.2 0 4.6-2.8 5.6-5.5 5.9.5.4.9 1 .9 2.2v3.3c0 .3.1.7.8.6A12 12 0 0 0 12 .3"
      />
    </svg>
  </span>
  <span
    class="MuiTouchRipple-root"
  />
</button>
`;

exports[`components Navigator snapshots contain navigations and its icon on light theme 3`] = `
<button
  class="MuiButtonBase-root MuiIconButton-root WhiteNavigator-themeContainer MuiIconButton-sizeSmall MuiIconButton-edgeEnd"
  tabindex="0"
  title="Light / Dark Theme"
  type="button"
>
  <span
    class="MuiIconButton-label"
  >
    <svg
      aria-hidden="true"
      class="MuiSvgIcon-root WhiteNavigator-compressIcon"
      focusable="false"
      role="presentation"
      viewBox="0 0 24 24"
    >
      <path
        d="M6.76 4.84l-1.8-1.79-1.41 1.41 1.79 1.79 1.42-1.41zM4 10.5H1v2h3v-2zm9-9.95h-2V3.5h2V.55zm7.45 3.91l-1.41-1.41-1.79 1.79 1.41 1.41 1.79-1.79zm-3.21 13.7l1.79 1.8 1.41-1.41-1.8-1.79-1.4 1.4zM20 10.5v2h3v-2h-3zm-8-5c-3.31 0-6 2.69-6 6s2.69 6 6 6 6-2.69 6-6-2.69-6-6-6zm-1 16.95h2V19.5h-2v2.95zm-7.45-3.91l1.41 1.41 1.79-1.8-1.41-1.41-1.79 1.8z"
      />
    </svg>
  </span>
  <span
    class="MuiTouchRipple-root"
  />
</button>
`;
